Fixed #29865 -- Added logical XOR support for Q() and QuerySet().

rheard opened the pull request:
Add logical XOR support to the ORM. MySQL will use the logical XOR, while other backends will convert

pope1ni commented on django/db/backends/oracle/features.py:
```diff
@@ -65,6 +65,7 @@ class DatabaseFeatures(BaseDatabaseFeatures):
     supports_primitives_in_json_field = False
     supports_json_field_contains = False
     supports_collation_on_textfield = False
+    supports_logical_xor = False
```
No need to specify this - it's already inherited from the base.

```suggestion
```

pope1ni commented on django/db/backends/postgresql/features.py:
```diff
@@ -58,6 +58,7 @@ class DatabaseFeatures(BaseDatabaseFeatures):
     supports_deferrable_unique_constraints = True
     has_json_operators = True
     json_key_contains_list_matching_requires_list = True
+    supports_logical_xor = False
```
No need to specify this - it's already inherited from the base.

```suggestion
```

pope1ni commented on django/db/backends/sqlite3/features.py:
```diff
@@ -44,6 +44,7 @@ class DatabaseFeatures(BaseDatabaseFeatures):
     supports_order_by_nulls_modifier = Database.sqlite_version_info >= (3, 30, 0)
     order_by_nulls_first = True
     supports_json_field_contains = False
+    supports_logical_xor = False
```
No need to specify this - it's already inherited from the base.

```suggestion
```

pope1ni commented on django/db/models/expressions.py:
```diff
@@ -105,14 +105,21 @@ def bitleftshift(self, other):
 
+    def __xor__(self, other):
+        if getattr(self, 'conditional', False) and getattr(other, 'conditional', False):
+            return Q(self) & Q(other)
```
Looks like you forgot to change the operator:

```suggestion
            return Q(self) ^ Q(other)
```

rheard replied:
Eeeek, good catch, looks like none of my unit tests hit that part. I'm not sure if Combinable/Expression/Subquery/Exists has an appropriate place for unit tests like, but I'm moving around some decorators and adding one.

If you have an idea for a better spot for `test_logical_xor_subquery`, let me know

pope1ni commented on django/db/models/expressions.py:
```diff
@@ -138,12 +145,12 @@ def __rpow__(self, other):
+            "Use .bitand(), .bitor(), and .bitxor() for bitwise logical operations."
         )
 
     def __ror__(self, other):
```
Please can you duplicate this for `__rxor__()`.

pope1ni commented on django/db/models/sql/where.py:
```diff
@@ -76,6 +77,9 @@ def as_sql(self, compiler, connection):
             full_needed, empty_needed = 1, len(self.children)
 
+        if self.connector == XOR and not connection.features.supports_logical_xor:
+            raise TypeError('The XOR operator is not supported on this database backend')
```
Please add a period:

```suggestion
            raise TypeError('The XOR operator is not supported on this database backend.')
```

pope1ni reviewed (changes requested):
Here's a very quick pass. I've not looked at the tests yet.

smithdc1 commented on docs/releases/4.0.txt:
```diff
@@ -249,6 +249,10 @@ Models
 * :class:`~django.db.models.DurationField` now supports multiplying and
   dividing by scalar values on SQLite.
 
+* :class:`Q() <django.db.models.Q>` now supports ``__xor__`` and will be
```
Please could you also add the new feature to the relevent documentation page(s).

rheard replied:
I have added documentation to the `QuerySet` and `Q` object documentations, but a review of the wording there may be needed. Thanks

pope1ni reviewed (changes requested):
Thanks for the updates.

Please update some of the existing tests.

Add the following to the end of `tests` in `QTests.test_combine_empty_copy`:

```python
            base_q ^ Q(),
            Q() ^ base_q,
```

Add the following to the end of `QTests.test_combine_not_q_object`:

```python
        with self.assertRaisesMessage(TypeError, str(obj)):
            q ^ obj
```

Add the following to the end of `tests` in `QTests.test_combine_negated_boolean_expression`:

```python
            Q() ^ ~Exists(tagged),
```

It seems as though we'd want to add to `QuerySetBitwiseOperationTests` too to test the `QuerySet` behavior?

Also, I don't think that we need to create `tests/xor_lookups`. These tests should fit in around `QuerySet`-related tests for `&` and `|`.

I've marked this as _patch needs improvement_ on ticket-29865. Please remove that flag when you're ready for another round of review. Thanks.

rheard commented:
> Thanks for the updates.
> 
> Please update some of the existing tests.
> 
> Add the following to the end of `tests` in `QTests.test_combine_empty_copy`:
> 
> ```python
>             base_q ^ Q(),
>             Q() ^ base_q,
> ```
> 
> Add the following to the end of `QTests.test_combine_not_q_object`:
> 
> ```python
>         with self.assertRaisesMessage(TypeError, str(obj)):
>             q ^ obj
> ```
> 
> Add the following to the end of `tests` in `QTests.test_combine_negated_boolean_expression`:
> 
> ```python
>             Q() ^ ~Exists(tagged),
> ```
> 
> It seems as though we'd want to add to `QuerySetBitwiseOperationTests` too to test the `QuerySet` behavior?
> 
> Also, I don't think that we need to create `tests/xor_lookups`. These tests should fit in around `QuerySet`-related tests for `&` and `|`.
> 
> I've marked this as _patch needs improvement_ on [ticket-29865](https://code.djangoproject.com/ticket/29865). Please remove that flag when you're ready for another round of review. Thanks.

I had some issues around rebasing and force pushing, a nightmare, I think it has been resolved. 

I've added those tests in and I added tests to `QuerySetBitwiseOperationTests` that emulate the "or" tests there, I'm not sure if its how you imagined though.

The `tests/xor_lookups` seemed appropriate as they were mostly copies of the existing `tests/or_lookups`. I can merge them both into `tests/or_lookups` perhaps? I'm just trying imagine the most logical places for them

pope1ni commented:
> The `tests/xor_lookups` seemed appropriate as they were mostly copies of the existing `tests/or_lookups`. I can merge them both into `tests/or_lookups` perhaps? I'm just trying imagine the most logical places for them

My mistake. That'll teach me for making this point based on the non-existent `test/and_lookups` and missing that `test/or_lookups` existed! 😆

pope1ni commented on docs/ref/models/querysets.txt:
```diff
@@ -1903,6 +1903,29 @@ SQL equivalent:
+
+    SELECT ... WHERE x=1 XOR y=2
+
+.. note::
+    ``XOR`` is emulated in backends that don't support it by changing
+    ``x=1 XOR y=2`` into ``(x=1 OR x=2) AND NOT (x=1 AND y=2)``.
```
Lets add a blank line after the directive and also make it clear which backends currently support this:
```suggestion
.. note::

    ``XOR`` is emulated in backends that don't support it by changing
    ``x=1 XOR y=2`` into ``(x=1 OR x=2) AND NOT (x=1 AND y=2)``. Currently
    ``XOR`` is only supported by MySQL and MariaDB.
```

rheard replied:
I have added this.

I only see/added backend settings for MySQL though, and I'm not sure if MariaDB uses those settings?

pope1ni commented on docs/ref/models/querysets.txt:
```diff
@@ -3712,8 +3735,12 @@ A ``Q()`` object represents an SQL condition that can be used in
+
+.. versionchanged:: 4.0
+
+    Support for the ``^`` (``XOR``) operator was added.
```
```suggestion
    Added support for the ``^`` (``XOR``) operator.
```

pope1ni commented on docs/releases/4.0.txt:
```diff
@@ -280,6 +280,11 @@ Models
 * :class:`~django.db.models.Lookup` expressions may now be used in ``QuerySet``
   annotations, aggregations, and directly in filters.
 
+* :class:`~django.db.models.Q` and :class:`~django.db.models.query.QuerySet`
+  now support ``__xor__``, and will be compiled on MySQL to logical ``XOR``.
+  Unsupported database backends will convert the query to an equivalent using
+  only ``AND``, ``OR``, and ``NOT``.
```
```suggestion
* :class:`~django.db.models.Q` and :class:`~django.db.models.query.QuerySet`
  can now be combined using ``^`` as the exclusive or (XOR) operator. For MySQL
  and MariaDB this is compiled to logical ``XOR``. For database backends that
  do not support ``XOR``, the query will be converted to an equivalent using
  only ``AND``, ``OR``, and ``NOT``.
```

pope1ni reviewed (commented):
I've suggested a few more tweaks in the documentation.

Please can you change the commit message and PR title to the following:

```
Fixed #29865 -- Added logical XOR support for Q() and QuerySet().
```

I've also noticed that the ticket was triaged and "Someday/Maybe" and not "Accepted". I'm sorry I didn't notice this before. However, based on the fact that we have what seems to be a working implementation here I'll change it to "Accepted". This will put it on the review queue for the fellows to follow up on.

pope1ni commented:
buildbot, test on oracle.

felixxm commented on docs/ref/models/querysets.txt:
```diff
@@ -1903,6 +1903,31 @@ SQL equivalent:
+
+.. note::
+
+    ``XOR`` is emulated in backends that don't support it by changing
+    ``x=1 XOR y=2`` into ``(x=1 OR x=2) AND NOT (x=1 AND y=2)``. Currently
+    ``XOR`` is only supported by MySQL and MariaDB.
```
Maybe:
```suggestion
    ``XOR`` is natively supported on MariaDB and MySQL. On other databases,
    this is emulated using ``(x=1 OR x=2) AND NOT (x=1 AND y=2)``.
```

felixxm commented on docs/releases/4.0.txt:
```diff
@@ -306,6 +306,12 @@ Models
+  can now be combined using ``^`` as the exclusive or (XOR) operator. For MySQL
+  and MariaDB this is compiled to logical ``XOR``. For database backends that
+  do not support ``XOR``, the query will be converted to an equivalent using
+  only ``AND``, ``OR``, and ``NOT``.
```
Please move it to the 4.1 release notes.

felixxm commented on tests/queries/tests.py:
```diff
@@ -2079,6 +2083,18 @@ def test_distinct_ordered_sliced_subquery(self):
         )
 
+    def test_logical_xor_subquery(self):
+        """Subquery should support logical XOR on supported databases"""
```
This docstring doesn't have much value, I'd remove it.

rheard replied:
I'm going to push back on this one a little:

I find it best to ALWAYS have a docstring explaining what the test is trying to test. This way the name and code don't need to be deciphered, and there is no confusion about the intended goal of the test.

Further, I see other tests throughout here with similar short docstrings?

felixxm replied:
We have many old tests in `queries` with unnecessary docstrings.

rheard replied:
I'm sorry, I just can't find it in me to make my tests less readable. I know you said you had some nitpicks you wanted to change before merging, perhaps you can do that?

Or can you suggest a more descriptive docstring for the test that explains what its testing?

felixxm commented on tests/queries/tests.py:
```diff
@@ -2079,6 +2083,18 @@ def test_distinct_ordered_sliced_subquery(self):
+            ),
+            [2],
+            attrgetter('id'),
+        )
```
I know that tests in `queries` are mixed up, however I would move it to the `Queries6Tests` class.

rheard replied:
Hm, ok. It seemed the most logical place to test the subquery was in `SubqueryTests`? 

I can't really tell what the intended goal of `Queries6Tests` is, but it doesn't matter much to me, so I can move it

felixxm commented on tests/xor_lookups/models.py:
```diff
@@ -0,0 +1,23 @@
+        return str(self.num)
+
+    class Meta:
+        ordering = ('num', )
```
Do we need a `Meta.ordering`?

rheard replied:
Probably not, I used models from another test folder.

I'm not ordering manually in the tests however, and I'm not sure what might happen on another backend that I can't test immediately. It doesn't change much, so I'm just going to leave it?

felixxm commented on tests/xor_lookups/tests.py:
```diff
@@ -0,0 +1,77 @@
+                5, 6, 7, 8, 9, 10,
+            ],
+            attrgetter('num'),
+        )
```
I have some nitpicks to these tests but I can push them later after the final review.

rheard replied:
Ok, they were mostly copied from `or_lookups`, so you may have a look there

felixxm reviewed (commented):
@rheard Thanks for this patch :+1:

rheard commented on django/db/models/sql/where.py:
```diff
@@ -76,6 +77,12 @@ def as_sql(self, compiler, connection):
         else:
             full_needed, empty_needed = 1, len(self.children)
 
+        if self.connector == XOR and not connection.features.supports_logical_xor:
+            # Convert `A XOR B` to `(A OR B) AND NOT (A AND B)`.
+            lhs = self.__class__(self.children, OR)
+            rhs = self.__class__(self.children, AND, negated=True)
+            return self.__class__((lhs, rhs), AND, self.negated).as_sql(compiler, connection)
```
@felixxm This is conversion is based on the feature found in `connection.features`.

From what I could tell, `Q` objects do not seem to have access to what DB they're intended for, and thus can't know if they should be converted or not.

That conversion work has to be done once we know the intended database backend, so here.

rheard commented on docs/ref/models/querysets.txt:
```diff
@@ -3728,8 +3753,12 @@ A ``Q()`` object represents an SQL condition that can be used in
+combine them using operators such as ``|`` (``OR``), ``&`` (``AND``),
+and ``^`` (``XOR``). See :ref:`complex-lookups-with-q`.
+
+.. versionchanged:: 4.0
+
+    Added support for the ``^`` (``XOR``) operator.
```
Whoops, apologies. There was some issues around rebasing. Fixing

felixxm commented:
@rheard Thanks for updates :+1:

rheard commented on django/db/models/sql/where.py:
```diff
@@ -42,7 +43,7 @@ def split_having(self, negated=False):
         may_need_split = (
             (in_negated and self.connector == AND) or
-            (not in_negated and self.connector == OR))
+            (not in_negated and self.connector in (OR, XOR)))
```
Tbh, this statement went passed my knowledge and I had no idea how to effectively handle it. I'm glad you're reviewing it

I can change it to what you've suggested. I'm not entirely sure the best way to test this, are there unit tests currently for this splitting?

felixxm replied:
I checked this carefully and when connector is `XOR` we always have to push the whole branch to `HAVING` clause.

felixxm commented on django/db/models/sql/where.py:
```diff
@@ -76,6 +77,12 @@ def as_sql(self, compiler, connection):
 
+        if self.connector == XOR and not connection.features.supports_logical_xor:
+            # Convert `A XOR B` to `(A OR B) AND NOT (A AND B)`.
+            lhs = self.__class__(self.children, OR)
+            rhs = self.__class__(self.children, AND, negated=True)
```
This doesn't work with many conditions, e.g. `a ^ b ^ c` see `XorLookupsTests.test_filter_negated()`.

rheard replied:
Great catch. 

It looks like 3,4,.. args would need some higher-level logic. I could accomplish that by going through the children in search of `XOR`s, and their children, and so on... and then building up equivalent logic from that list

I will work on that tonight when I have free time after work, it should not be too difficult. Though I'm already stressing about how I'm going to keep it clean.

rheard replied:
It was easier than I thought. instead of `(A OR B) AND NOT (A AND B)`, the new logic is:
`(A OR B OR ...) AND 1 = (CASE(1 if A else 0) + CASE(1 if B else 0) + ...)`

NOTE: I tried to use `Cast` or `ExpressionWrapper` instead but they did not seem to work

felixxm reviewed (commented):
@rheard I pushed small edits to tests and docs. Unfortunately, I found an issue in chaining multiple conditions. Can you take a look? 

Also, please don't rebase this patch and push edits in separate commits. I will squash everything before merging.

rheard commented:
> @rheard I pushed small edits to tests and docs. Unfortunately, I found an issue in chaining multiple conditions. Can you take a look?
> 
> Also, please don't rebase this patch and push edits in separate commits. I will squash everything before merging.

Not a problem! I proposed a solution and will begin working on it tonight.

It looks like the ordering was removed from the `Meta` though and broke some tests :/

felixxm commented:
> It looks like the ordering was removed from the `Meta` though and broke some tests :/

I changed to `assertSequenceEqual()` and `assertCountEqual()` so `Meta.ordering` is not necessary anymore. Only the regression test (`xor_lookups.tests.XorLookupsTests.test_filter_negated`) for many conditions is broken on SQLite and PostgreSQL (as expected) .

rheard commented:
> > It looks like the ordering was removed from the `Meta` though and broke some tests :/
> 
> I changed to `assertSequenceEqual()` and `assertCountEqual()` so `Meta.ordering` is not necessary anymore. Only the regression test (`xor_lookups.tests.XorLookupsTests.test_filter_negated`) for many conditions is broken on SQLite and PostgreSQL (as expected) .

Sorry, quickly realized that. Again, will implement my proposed solution when I have time. Thanks!

kezabelle commented on django/db/models/sql/where.py:
```diff
@@ -41,7 +44,7 @@ def split_having(self, negated=False):
         # then we need to push the whole branch to HAVING clause.
         may_need_split = (
-            (in_negated and self.connector == AND) or
+            (in_negated and self.connector == (AND, XOR)) or
```
Flagging this because I think the change was made based on Mariusz' suggestion (somewhere upthread), but the operator has accidentally been left unchanged. This should be `in` rather than `==` I think?

rheard replied:
Oof, ouch. Thanks for that. Fixed

felixxm commented:
I rebased from the `main` branch, resolved conflicts, and applied `black`.

felixxm reviewed (approved):
@rheard Thanks for all your efforts :+1: Welcome aboard :boat: 

I pushed final edits and squashed commits.

felixxm commented:
buildbot, test on oracle.
